### PR TITLE
feat(proxy): stop idle proxy-started daemons

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/pitchfork/trails/18

Opening a proxy URL starts a stopped daemon and, since #891, its `depends`. But nothing ever stopped them again: a dev server opened once in the browser kept running, with its database, until stopped by hand. This adds opt-in idle shutdown for daemons the proxy started. It is **off by default**, so existing setups behave exactly as before.

```toml
# ~/.config/pitchfork/config.toml, or a project's pitchfork.toml
[settings.proxy]
idle_timeout = "15m"            # every proxy-started daemon (default: "" = off)

[daemons.web]
run = "npm run dev"
port = 5173
depends = ["db"]
proxy_idle_timeout = "5m"       # per-daemon override; `false` opts out
```

With this, opening `web.<project>.localhost` starts `db` and `web`. Five minutes after the proxy last carried anything for `web`, pitchfork stops `web`, then `db`, which nothing running needs any more. The next visit starts both again. The setting can also come from the environment (`PITCHFORK_PROXY_IDLE_TIMEOUT`).

**Which daemons can be stopped (ownership)**
- Only daemons the proxy auto-started, plus the dependencies it started along with them. A dependency without its own `proxy_idle_timeout` uses the value of the daemon that pulled it in.
- Anything started another way is never idle-stopped: `pitchfork start`/`restart`/`run`, the TUI, the web UI, the shell hook, `boot_start`.
- An explicit start of a daemon the proxy already started claims that daemon and its dependencies, and they then stay up until stopped. This applies from the CLI, TUI, MCP and web UI. The outcome is the same whichever came first. If the claim fails, the start fails with it rather than leaving the daemon idle-stoppable. Only a supervisor too old to know the request is tolerated.
- Retries and file-watch restarts keep the daemon's ownership. It is stored in the state file, so it survives a supervisor restart.

**What counts as activity.** HTTP/HTTPS traffic through the proxy, never DNS lookups:
- a request counts until its response body has been fully sent, so SSE and other streaming responses count for as long as they stream;
- WebSockets and other upgraded connections count while open, even when quiet, so a tab with an HMR socket keeps its dev server alive;
- TLS passthrough connections count while open;
- idle keep-alive connections don't count.

**What keeps a daemon running**
- **Dependents:** a daemon depending on it is running, starting, or on its way back (stopping for a restart, or errored with retries left). Shared dependencies stay while any consumer runs, and daemons stop in reverse dependency order.
- **Shell sessions:** a tracked shell or `project enter` session is inside its directory.
- **Workers and databases:** non-proxied daemons have no traffic of their own, so they're never judged idle on it. They go only as proxy-started dependencies, after their last dependent.

**Races.** A single lock both counts activity and claims a stop:
- A stop can only be claimed with nothing in flight.
- A request that arrives while a stop is under way waits for it to finish (bounded by `proxy.auto_start_timeout`), then starts the daemon again in the same request.
- Proxy auto-start holds activity on the daemon's whole dependency graph for as long as its start runs. That covers #891's background continuation after the request's timeout, so a dependency it is about to rely on can't be stopped under it.
- Everything is re-checked under the daemon's stop lock right before stopping: still running, still proxy-owned, no dependents, no shell.
- An explicit start's claim waits out an idle stop already in progress.

**Timing and limitations** (documented in the proxy guide, *Idle Shutdown*)
- Idleness is checked every `general.interval` (default 10s), so a daemon stops between the grace period and the grace period plus one interval.
- Only proxy traffic is visible. A client hitting the daemon's port directly isn't seen; start such daemons explicitly, or send a periodic request through the proxy URL as a heartbeat.
- The grace period is fixed when the proxy starts the daemon.
- After a supervisor restart, each proxy-started daemon gets a fresh grace period.

**Validation**
- New `test/idle.bats` lifecycle tests:
  - stop order (dependent before dependency), with an explicit daemon left alone and a restart on the next request;
  - off by default, plus the per-daemon opt-in;
  - a long SSE stream and a quiet WebSocket keep the daemon alive;
  - an explicit start claims the daemon;
  - a shell in the project vetoes the stop until it leaves;
  - a shared dependency stays for a remaining consumer.
- I checked that the streaming, WebSocket, claim and shell tests each fail when their safeguard is removed.
- Unit tests cover the stop planner (ordering, shared dependencies, starting dependents, shells, cycles), the activity tracker's claim/refuse behavior, the config parsing, and IPC round-trips of the new fields and request.
- `mise run ci-dev` passes locally on Linux.

*AI-assisted — Tool: Claude Code; model: Anthropic/claude-opus-5; version: 2.1.270.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches proxy hot path, supervisor lifecycle, and IPC with race-sensitive idle-stop/claim logic; default-off config limits blast radius but mis-tuned timeouts could surprise dev workflows.
> 
> **Overview**
> Adds **opt-in idle shutdown** for daemons the reverse proxy auto-starts. Global `proxy.idle_timeout` / `PITCHFORK_PROXY_IDLE_TIMEOUT` and per-daemon `proxy_idle_timeout` control grace periods; both default to off so existing setups are unchanged.
> 
> The proxy tracks **activity** (HTTP until the body finishes, WebSockets, TLS passthrough, auto-start graphs) and the supervisor periodically stops eligible daemons in **dependent-first** order when nothing in flight, no live dependents, and no tracked shell/project session is in the daemon directory. Explicit starts send **`ClaimDaemons`** IPC to clear proxy ownership (including dependencies); proxy auto-starts stamp `proxy_idle_timeout_ms` on state and `RunOptions`.
> 
> Requests that hit a daemon mid idle-stop **wait** (within `proxy.auto_start_timeout`) and restart it; streaming responses use guarded bodies so long streams count as active. Docs, JSON schema, CLI metadata, and `test/idle.bats` cover configuration and lifecycle edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c7f4810d3e7934003f60af1816cf1bbc4fd85541. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **New Features**
  - Added optional idle shutdown for daemons automatically started through the proxy; disabled by default.
  - Added global and per-daemon timeout configuration, including opt-out support.
  - Active requests, streams, WebSockets, TLS connections, shell sessions, and dependent daemons prevent shutdown.
  - Explicitly started daemons remain running and restart automatically when accessed.
  - Idle checks run periodically, stopping dependencies after dependents.

- **Documentation**
  - Added configuration, CLI, schema, and port-management documentation for idle shutdown.

### CODE REVIEW

File: src/ipc/client.rs [hunk 1]
```diff
@@ -584,6 +584,25 @@ impl IpcClient {
...
+        if ids.is_empty() {
+            return;
+        }
+        match self
+            .request(IpcRequest::ClaimDaemons { ids: ids.to_vec() })
+            .await
+        {
+            Ok(IpcResponse::Ok) => {}
+            Ok(rsp) => debug!("supervisor did not accept ClaimDaemons: {rsp:?}"),
+            Err(e) => debug!("failed to send ClaimDaemons: {e}"),
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/ipc/client.rs]: Valid, fixed in e8f14b4. `claim_daemons` now returns a `Result`, and every explicit start path propagates it (`start_daemons_with_config`, `start_daemon`, `run_adhoc`). If the claim fails, the start fails instead of skipping a running proxy-owned daemon. The one refusal tolerated is an older supervisor that can't decode the new request variant: it answers `Invalid request: …`, and has no idle shutdown to opt out of.

*AI-assisted — Tool: Claude Code; model: Anthropic/claude-opus-5; version: 2.1.270.*

File: src/supervisor/idle.rs [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+/// A daemon's idle grace period, if it may be stopped for inactivity.
+fn grace(daemon: &Daemon) -> Option<Duration> {
+    daemon.proxy_idle_timeout_ms.map(Duration::from_millis)
+}
+
+/// Whether `daemon` is up or on its way up, and so still needs what it
+/// depends on.
+fn is_live(daemon: &Daemon) -> bool {
+    daemon.status.is_running() || daemon.status.is_waiting()
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs:85]: Valid, fixed in e8f14b4. A dependent now keeps its dependencies when it is running, waiting, **stopping** (a restart stops first), or **errored with retries left**. That's the same retries-remaining test `check_retry` uses. Unit test `a_dependent_on_its_way_back_keeps_its_dependency` covers stopping, errored with a retry pending, and errored with retries exhausted (which no longer protects). The final re-check under the dependency's stop lock uses the same predicate.

*AI-assisted — Tool: Claude Code; model: Anthropic/claude-opus-5; version: 2.1.270.*

File: src/ipc/batch.rs [hunk 3]
```diff
@@ -1133,6 +1188,9 @@ impl IpcClient {
             .await
             .map_err(|e| miette::miette!("{e}"))?;
 
+        if overrides.is_none_or(|o| o.proxy_idle.is_none()) {
+            self.claim_daemons(std::slice::from_ref(id)).await;
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/ipc/batch.rs:1206]: Valid, fixed in e8f14b4. `start_daemon` (web UI start/restart) now claims the daemon's full dependency closure before starting, exactly like `start_daemons`.

*AI-assisted — Tool: Claude Code; model: Anthropic/claude-opus-5; version: 2.1.270.*

File: src/proxy/server.rs [hunk 4]
```diff
@@ -3325,13 +3345,86 @@ async fn resolve_target(host: &str, tld: &str) -> ResolveResult {
...
+
+/// Record a request for a daemon that the state showed running, then check
+/// that it still is.
+///
+/// Recording comes first: once it has, a stop for inactivity can no longer be
+/// claimed, so a daemon that is still running at the check stays running for
+/// as long as the request holds the guard.
+async fn begin_running(id: &DaemonId) -> Running {
+    let Some(activity) = ACTIVITY.begin(id) else {
+        return if wait_out_idle_stops(std::slice::from_ref(id)).await {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/proxy/server.rs]: Fixed in fde6303: `resolve_target` and the registry path now compute one `proxy.auto_start_timeout` deadline and pass it to both the idle-stop wait and `try_auto_start`, so the whole request stays within one timeout.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/proxy/server.rs [hunk 5]
```diff
@@ -3325,13 +3345,86 @@ async fn resolve_target(host: &str, tld: &str) -> ResolveResult {
...
+            Running::No
+        } else {
+            Running::Stopping
+        };
+    };
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/proxy/server.rs]: Fixed in fde6303: after the claim clears, `begin_running` retries `ACTIVITY.begin` and re-reads the state, so a called-off stop forwards to the still-running daemon (new test `a_request_forwards_after_an_idle_stop_is_called_off`).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+                    debug!("idle stop of {id} called off: {reason}");
+                    false
+                }
+                None => {
+                    info!("stopping {id}: no proxy activity for {grace:?}");
+                    match self.stop_locked(id).await {
+                        Ok(_) => true,
+                        Err(e) => {
+                            error!("failed to stop idle daemon {id}: {e}");
+                            false
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs]: Fixed in fde6303: only `Ok`, `DaemonWasNotRunning` and `DaemonNotRunning` count as stopped; any other response (e.g. `DaemonStopFailed`) is logged as a failure and no notification is sent.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/scripts/hold_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+Usage: hold_upgrade.py <proxy_port> <host> <path> <seconds>
+"""
+import socket
+import sys
+import time
+
+proxy_port, host, path, secs = int(sys.argv[1]), sys.argv[2], sys.argv[3], float(sys.argv[4])
+
+sock = socket.create_connection(("127.0.0.1", proxy_port))
```
[the remark this answers is not in the document]
Comment (jdx) [reply test/scripts/hold_upgrade.py]: Fixed in fde6303: the handshake socket now has a 20s timeout, so a proxy that never answers fails the test instead of hanging it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/proxy/server.rs [hunk 8]
```diff
@@ -3833,8 +3950,15 @@ async fn wait_for_active_port(daemon_id: &DaemonId, route: &ProxyTlsRoute) -> Re
...
                 // later ones.
                 if let Some(port) = select_daemon_port(route, d) {
-                    log::info!("Auto-start: daemon {daemon_id} is ready on port {port}");
-                    return ResolveResult::Ready(port);
+                    // Recorded as the request's activity. Its start held the
+                    // daemon's activity until it finished, so a stop for
+                    // inactivity can only have been claimed since if the grace
+                    // period is shorter than this poll; then the next poll
+                    // finds it stopping.
+                    if let Some(activity) = ACTIVITY.begin(daemon_id) {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/proxy/server.rs]: Fixed in fde6303: a `None` from `ACTIVITY.begin` in `wait_for_active_port` now returns `Starting`. The page reloads (and passthrough resolves again), waits out the stop, and starts the daemon again within the timeout.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+                !chosen
+                    .iter()
+                    .any(|other| daemons.get(other).is_some_and(|d| d.depends.contains(id)))
+            })
+            .cloned()
+            .collect();
+        if level.is_empty() {
+            // A dependency cycle: nothing can go first, so stop them together.
+            level = chosen.iter().cloned().collect();
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs:163]: Fixed in fde6303: each plan level is claimed as a group, and members are validated with claimed peers set aside; one that no longer qualifies (or fails to stop) leaves the group so peers keep what it needs. (`depends` cycles are rejected at start, so this is defensive.)

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+                    }
+                }
+            };
+            if stopped {
+                self.add_notification(Info, format!("stopped idle {id}"))
+                    .await;
+            } else {
+                // Still running, so the members checked after it must keep
+                // what it needs.
+                group.remove(&id);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs:340]: Fixed in 11f030c: the planner no longer stops cycle members together. It stops planning at a cycle and leaves the cycle and what it depends on running, since a partial cycle stop can't be undone, and `depends` cycles are refused at start anyway. Test `a_dependency_cycle_is_left_running`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+            }
+        }
+
+        for id in &claimed {
+            ACTIVITY.release_idle_stop(id);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs]: Fixed in 11f030c: a member that leaves the level group has its idle-stop claim released immediately, and every other member's claim is released right after it's handled, instead of at the end of the level.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/ipc/batch.rs [hunk 12]
```diff
@@ -1315,6 +1350,8 @@ impl IpcClient {
             );
         }
 
+        self.claim_daemons(std::slice::from_ref(&run_opts.id))
+            .await?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/ipc/batch.rs:1360]: Fixed in 40b8120: `run_adhoc` now claims the configured daemon's full dependency closure (falling back to just the id for a daemon not in config), matching `start_daemons` and `start_daemon`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+    /// Why `id` must not be stopped for inactivity right now, if anything.
+    ///
+    /// Dependents in `stopping_with` are being stopped along with it and do
+    /// not count.
+    async fn idle_stop_blocker(
+        &self,
+        id: &DaemonId,
+        stopping_with: &HashSet<DaemonId>,
+    ) -> Option<&'static str> {
+        let active_dirs = self.get_active_directories().await;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs]: Fixed in 40b8120: the final check now reads shells and project sessions under the same state lock as the rest of its checks (`StateFile::active_directories`), so no shell can enter between the snapshot and the check. A shell that enters after the final check behaves like one entering just after the stop: the shell hook starts the daemon if it's configured to, so I didn't add locking to the shell-hook handlers.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/ipc/batch.rs [hunk 14]
```diff
@@ -1318,6 +1353,19 @@ impl IpcClient {
...
+        // so nothing else would keep a proxy-started dependency it may be
+        // using from being stopped when idle.
+        let claimed: Vec<DaemonId> = PitchforkToml::all_merged_all_namespaces()
+            .ok()
+            .filter(|pt| pt.daemons.contains_key(&run_opts.id))
+            .and_then(|pt| {
+                resolve_dependencies(std::slice::from_ref(&run_opts.id), &pt.daemons).ok()
+            })
+            .map(|order| order.levels.into_iter().flatten().collect())
+            .unwrap_or_else(|| vec![run_opts.id.clone()]);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/ipc/batch.rs]: Agreed, reverted in 4d4168a: `run_adhoc` claims only its own ID again. An ad-hoc command starts no dependencies and declares no `depends`, so claiming a configured daemon's graph would wrongly make unrelated daemons like `db` explicit.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+    /// not count.
+    async fn idle_stop_blocker(
+        &self,
+        id: &DaemonId,
+        stopping_with: &HashSet<DaemonId>,
+    ) -> Option<&'static str> {
+        // Shells and sessions are read under the same lock as the rest, so a
+        // shell entering the directory cannot slip in between the two.
+        let state_file = self.state_file.lock().await;
+        let active_dirs = state_file.active_directories();
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs:374]: Fixed in 4d4168a: a shared shell-admission lock is held by the idle stop from its final check until `stop_locked` returns, and briefly by `UpdateShellDir`/`ProjectEnter` around registering the directory. A shell is either seen by the check or registered after the daemon has stopped, the same as entering a moment later. Entering a directory waits only while an idle stop is running.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/ipc/batch.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jdx) [reply src/ipc/batch.rs:1360]: Follow-up: reverted in 4d4168a after review of #discussion_r4054701280. An ad-hoc `pitchfork run` executes its own command and starts no dependencies, so claiming the configured daemon's dependency graph would permanently take unrelated daemons out of idle shutdown. It now claims only its own ID.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/supervisor/idle.rs [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+        }
+
+        let mut ordered: Vec<DaemonId> = group.iter().cloned().collect();
+        ordered.sort();
+        for id in ordered {
+            let lock = self.stop_lock(&id).await;
+            let stopped = {
+                // Admission first, then the daemon's stop lock, as nothing
+                // takes them the other way round.
+                let _admission = SHELL_ADMISSION.lock().await;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/supervisor/idle.rs]: Fixed in c7f4810: the admission lock now covers only the final check plus marking the daemon `Stopping`, done as one step under the state lock. It's released before `stop_locked`, so a `cd` or `project enter` never waits for the daemon to exit. A shell entering afterwards finds the daemon `Stopping` rather than running, so the shell hook's start waits on the stop lock and brings it back. If the stop leaves the mark unsettled (no PID, or a failure), it's replaced with `Stopped` or `Running`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
